File: PATTERNS/fast_slow_pointers.py
"""
FAST & SLOW POINTERS (FLOYD'S CYCLE DETECTION)
Use: Cycle detection, finding middle, linked list problems
Time: O(n), Space: O(1)
"""

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# === CYCLE DETECTION ===
def has_cycle(head: Optional[ListNode]) -> bool:
    """Detect if linked list has a cycle"""
    if not head or not head.next:
        return False

    slow = head
    fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

        if slow == fast:
            return True

    return False


def find_cycle_start(head: Optional[ListNode]) -> Optional[ListNode]:
    """Find the node where cycle begins"""
    if not head or not head.next:
        return None

    slow = fast = head

    # detect cycle
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

        if slow == fast:
            break
    else:
        return None  # no cycle

    # find start: move slow to head, advance both by 1
    slow = head
    while slow != fast:
        slow = slow.next
        fast = fast.next

    return slow


def cycle_length(head: Optional[ListNode]) -> int:
    """Find length of cycle (0 if no cycle)"""
    if not head or not head.next:
        return 0

    slow = fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

        if slow == fast:
            # count cycle length
            length = 1
            current = slow.next
            while current != slow:
                length += 1
                current = current.next
            return length

    return 0


# === FIND MIDDLE ===
def find_middle(head: Optional[ListNode]) -> Optional[ListNode]:
    """Find middle node (second middle if even length)"""
    if not head:
        return None

    slow = fast = head

    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next

    return slow


def find_middle_first(head: Optional[ListNode]) -> Optional[ListNode]:
    """Find middle node (first middle if even length)"""
    if not head:
        return None

    slow = fast = head

    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    return slow


# === HAPPY NUMBER ===
def is_happy_number(n: int) -> bool:
    """Check if number is happy (sum of squares reaches 1)"""
    def get_next(num: int) -> int:
        total = 0
        while num:
            digit = num % 10
            total += digit * digit
            num //= 10
        return total

    slow = n
    fast = get_next(n)

    while fast != 1 and slow != fast:
        slow = get_next(slow)
        fast = get_next(get_next(fast))

    return fast == 1


# === FIND DUPLICATE (array as linked list) ===
def find_duplicate(nums: list) -> int:
    """
    Find duplicate in array where nums[i] in [1, n].
    Treat array as linked list: i -> nums[i]
    """
    # phase 1: find intersection
    slow = fast = nums[0]

    while True:
        slow = nums[slow]
        fast = nums[nums[fast]]
        if slow == fast:
            break

    # phase 2: find entrance
    slow = nums[0]
    while slow != fast:
        slow = nums[slow]
        fast = nums[fast]

    return slow


# === PALINDROME CHECK ===
def is_palindrome(head: Optional[ListNode]) -> bool:
    """Check if linked list is palindrome using O(1) space"""
    if not head or not head.next:
        return True

    # find middle
    slow = fast = head
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    # reverse second half
    second_half = reverse_list(slow.next)

    # compare halves
    first = head
    second = second_half
    result = True

    while second:
        if first.val != second.val:
            result = False
            break
        first = first.next
        second = second.next

    # restore list (optional)
    slow.next = reverse_list(second_half)

    return result


def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    """Helper: reverse linked list"""
    prev = None
    curr = head

    while curr:
        next_node = curr.next
        curr.next = prev
        prev = curr
        curr = next_node

    return prev


# === REORDER LIST ===
def reorder_list(head: Optional[ListNode]) -> None:
    """Reorder: L0 -> Ln -> L1 -> Ln-1 -> L2 -> Ln-2 -> ..."""
    if not head or not head.next:
        return

    # find middle
    slow = fast = head
    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    # reverse second half
    second = reverse_list(slow.next)
    slow.next = None
    first = head

    # merge alternating
    while second:
        tmp1, tmp2 = first.next, second.next
        first.next = second
        second.next = tmp1
        first = tmp1
        second = tmp2


# === NTH NODE FROM END ===
def remove_nth_from_end(head: Optional[ListNode], n: int) -> Optional[ListNode]:
    """Remove nth node from end using two pointers"""
    dummy = ListNode(0, head)
    slow = fast = dummy

    # advance fast by n+1
    for _ in range(n + 1):
        fast = fast.next

    # advance both until fast reaches end
    while fast:
        slow = slow.next
        fast = fast.next

    # remove node
    slow.next = slow.next.next

    return dummy.next


# === CIRCULAR ARRAY LOOP ===
def circular_array_loop(nums: list) -> bool:
    """Detect cycle in circular array (all same direction)"""
    n = len(nums)

    def next_idx(i: int) -> int:
        return (i + nums[i]) % n

    for i in range(n):
        if nums[i] == 0:
            continue

        slow = fast = i
        # ensure same direction
        while nums[slow] * nums[next_idx(slow)] > 0 and \
              nums[fast] * nums[next_idx(fast)] > 0 and \
              nums[next_idx(fast)] * nums[next_idx(next_idx(fast))] > 0:

            slow = next_idx(slow)
            fast = next_idx(next_idx(fast))

            if slow == fast:
                # check cycle length > 1
                if slow == next_idx(slow):
                    break
                return True

        # mark visited nodes
        j = i
        while nums[j] * nums[next_idx(j)] > 0:
            temp = j
            j = next_idx(j)
            nums[temp] = 0

    return False

File: PATTERNS/fast_slow_pointers.cpp
/*
 * FAST & SLOW POINTERS (FLOYD'S CYCLE DETECTION)
 * Use: Cycle detection, finding middle, linked list problems
 * Time: O(n), Space: O(1)
 */

#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

// === CYCLE DETECTION ===
bool hasCycle(ListNode* head) {
    if (!head || !head->next) return false;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) {
            return true;
        }
    }

    return false;
}

ListNode* findCycleStart(ListNode* head) {
    if (!head || !head->next) return nullptr;

    ListNode* slow = head;
    ListNode* fast = head;

    // detect cycle
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) {
            break;
        }
    }

    if (!fast || !fast->next) return nullptr;  // no cycle

    // find start
    slow = head;
    while (slow != fast) {
        slow = slow->next;
        fast = fast->next;
    }

    return slow;
}

int cycleLength(ListNode* head) {
    if (!head || !head->next) return 0;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;

        if (slow == fast) {
            int length = 1;
            ListNode* curr = slow->next;
            while (curr != slow) {
                length++;
                curr = curr->next;
            }
            return length;
        }
    }

    return 0;
}

// === FIND MIDDLE ===
ListNode* findMiddle(ListNode* head) {
    if (!head) return nullptr;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow;
}

ListNode* findMiddleFirst(ListNode* head) {
    if (!head) return nullptr;

    ListNode* slow = head;
    ListNode* fast = head;

    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow;
}

// === HAPPY NUMBER ===
int getNext(int num) {
    int total = 0;
    while (num) {
        int digit = num % 10;
        total += digit * digit;
        num /= 10;
    }
    return total;
}

bool isHappyNumber(int n) {
    int slow = n;
    int fast = getNext(n);

    while (fast != 1 && slow != fast) {
        slow = getNext(slow);
        fast = getNext(getNext(fast));
    }

    return fast == 1;
}

// === FIND DUPLICATE (array as linked list) ===
int findDuplicate(vector<int>& nums) {
    // phase 1: find intersection
    int slow = nums[0];
    int fast = nums[0];

    do {
        slow = nums[slow];
        fast = nums[nums[fast]];
    } while (slow != fast);

    // phase 2: find entrance
    slow = nums[0];
    while (slow != fast) {
        slow = nums[slow];
        fast = nums[fast];
    }

    return slow;
}

// === REVERSE LIST HELPER ===
ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr;
    ListNode* curr = head;

    while (curr) {
        ListNode* next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }

    return prev;
}

// === PALINDROME CHECK ===
bool isPalindrome(ListNode* head) {
    if (!head || !head->next) return true;

    // find middle
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    // reverse second half
    ListNode* secondHalf = reverseList(slow->next);

    // compare halves
    ListNode* first = head;
    ListNode* second = secondHalf;
    bool result = true;

    while (second) {
        if (first->val != second->val) {
            result = false;
            break;
        }
        first = first->next;
        second = second->next;
    }

    // restore list (optional)
    slow->next = reverseList(secondHalf);

    return result;
}

// === REORDER LIST ===
void reorderList(ListNode* head) {
    if (!head || !head->next) return;

    // find middle
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast->next && fast->next->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    // reverse second half
    ListNode* second = reverseList(slow->next);
    slow->next = nullptr;
    ListNode* first = head;

    // merge alternating
    while (second) {
        ListNode* tmp1 = first->next;
        ListNode* tmp2 = second->next;
        first->next = second;
        second->next = tmp1;
        first = tmp1;
        second = tmp2;
    }
}

// === NTH NODE FROM END ===
ListNode* removeNthFromEnd(ListNode* head, int n) {
    ListNode dummy(0);
    dummy.next = head;
    ListNode* slow = &dummy;
    ListNode* fast = &dummy;

    // advance fast by n+1
    for (int i = 0; i <= n; i++) {
        fast = fast->next;
    }

    // advance both until fast reaches end
    while (fast) {
        slow = slow->next;
        fast = fast->next;
    }

    // remove node
    ListNode* toDelete = slow->next;
    slow->next = slow->next->next;
    delete toDelete;

    return dummy.next;
}

// === CIRCULAR ARRAY LOOP ===
bool circularArrayLoop(vector<int>& nums) {
    int n = nums.size();

    auto nextIdx = [&](int i) {
        return ((i + nums[i]) % n + n) % n;
    };

    for (int i = 0; i < n; i++) {
        if (nums[i] == 0) continue;

        int slow = i, fast = i;
        // ensure same direction
        while (nums[slow] * nums[nextIdx(slow)] > 0 &&
               nums[fast] * nums[nextIdx(fast)] > 0 &&
               nums[nextIdx(fast)] * nums[nextIdx(nextIdx(fast))] > 0) {

            slow = nextIdx(slow);
            fast = nextIdx(nextIdx(fast));

            if (slow == fast) {
                // check cycle length > 1
                if (slow == nextIdx(slow)) break;
                return true;
            }
        }

        // mark visited nodes
        int j = i;
        int sign = nums[i];
        while (nums[j] * sign > 0) {
            int next = nextIdx(j);
            nums[j] = 0;
            j = next;
        }
    }

    return false;
}
